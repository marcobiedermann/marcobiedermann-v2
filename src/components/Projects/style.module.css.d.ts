/* eslint-disable import/prefer-default-export */

export const projects: string;
