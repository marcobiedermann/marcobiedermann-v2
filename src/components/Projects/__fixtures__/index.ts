import project from '../../Project/__fixtures__';

const projects = [project];

export default projects;
