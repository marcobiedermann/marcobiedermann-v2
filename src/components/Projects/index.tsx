import classNames from 'classnames';
import React from 'react';
import Project, { ProjectProps } from '../Project';
import styles from './style.module.css';

export interface ProjectsProps {
  className?: string;
  projects: ProjectProps[];
}

const Projects: React.FC<ProjectsProps> = (props) => {
  const { className, projects, ...otherProps } = props;

  return (
    <ol className={classNames(className, styles.projects)} {...otherProps}>
      {projects.map((project) => (
        <li key={project.id}>
          <Project {...project} />
        </li>
      ))}
    </ol>
  );
};

export default Projects;
