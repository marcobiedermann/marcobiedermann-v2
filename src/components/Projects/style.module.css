.projects {
  display: grid;
  grid-auto-flow: dense;
  grid-gap: 1em;
  grid-template-columns: repeat(6, 1fr);
  list-style: none;
  margin: 0;
  padding: 0;
}

.projects > li {
  grid-column-end: span 3;
}

.projects > li:nth-child(5n + 1) {
  grid-column-end: span 4;
  grid-row-end: span 2;
}

.projects > li:nth-child(5n + 2),
.projects > li:nth-child(5n + 3) {
  grid-column-end: span 2;
}
