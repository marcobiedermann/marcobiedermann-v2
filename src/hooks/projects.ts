import { graphql, useStaticQuery } from 'gatsby';
import { ProjectProps } from '../components/Project';

export interface ProjectsPageQuery {
  allContentfulProject: {
    projects: [
      {
        next: {
          id: string;
          slug: string;
        };
        node: ProjectProps;
        previous: {
          id: string;
          slug: string;
        };
      },
    ];
    pageInfo: {
      currentPage: number;
      hasNextPage: boolean;
      hasPreviousPage: boolean;
      pageCount: number;
    };
  };
}

export const useProjectsPage = (): ProjectsPageQuery => {
  return useStaticQuery<ProjectsPageQuery>(
    graphql`
      query {
        allContentfulProject(sort: { fields: [date], order: DESC }) {
          projects: edges {
            next {
              id
              slug
            }
            node {
              id
              slug
              title
              thumbnail {
                fluid {
                  ...GatsbyContentfulFluid_withWebp_noBase64
                }
              }
            }
            previous {
              id
              slug
            }
          }
        }
      }
    `,
  );
};
