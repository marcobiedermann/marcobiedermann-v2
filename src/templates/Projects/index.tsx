import { graphql } from 'gatsby';
import React from 'react';
import { useTranslation } from 'react-i18next';
import Grid from '../../components/Grid';
import Layout from '../../components/Layout';
import Pagination from '../../components/Pagination';
import { ProjectProps } from '../../components/Project';
import Projects from '../../components/Projects';
import Section from '../../components/Section';

export interface ProjectsTemplate {
  data: {
    allContentfulProject: {
      edges: [
        {
          next: {
            id: string;
            slug: string;
          };
          node: ProjectProps;
          previous: {
            id: string;
            slug: string;
          };
        },
      ];
      pageInfo: {
        currentPage: number;
        hasNextPage: boolean;
        hasPreviousPage: boolean;
        pageCount: number;
      };
    };
  };
}

const ProjectsTemplate: React.FC<ProjectsTemplate> = (props) => {
  const {
    data: { allContentfulProject },
  } = props;
  const { t } = useTranslation();

  const { edges: projects, pageInfo } = allContentfulProject;

  return (
    <Layout>
      <Section>
        <Grid>
          <Projects projects={projects.map((project) => project.node)} />
          <Pagination
            {...(pageInfo.hasNextPage && {
              next: {
                id: 'next',
                name: t('pagination.next'),
                path: `/projects/${pageInfo.currentPage + 1}`,
              },
            })}
            {...(pageInfo.hasPreviousPage && {
              previous: {
                id: 'previous',
                name: t('pagination.previous'),
                path: pageInfo.currentPage === 2 ? '/projects' : `/projects/${pageInfo.currentPage - 1}`,
              },
            })}
            routes={Array.from({ length: pageInfo.pageCount }, (_, index) => ({
              id: `${index}`,
              name: `${index + 1}`,
              path: index === 0 ? '/projects' : `/projects/${index + 1}`,
            }))}
          />
        </Grid>
      </Section>
    </Layout>
  );
};

export const pageQuery = graphql`
  query Projects($limit: Int!, $skip: Int!) {
    allContentfulProject(sort: { fields: [date], order: DESC }, limit: $limit, skip: $skip) {
      edges {
        next {
          id
          slug
        }
        node {
          id
          slug
          title
          thumbnail {
            fluid {
              ...GatsbyContentfulFluid_withWebp_noBase64
            }
          }
        }
        previous {
          id
          slug
        }
      }
      pageInfo {
        currentPage
        hasNextPage
        hasPreviousPage
        pageCount
      }
    }
  }
`;

export default ProjectsTemplate;
