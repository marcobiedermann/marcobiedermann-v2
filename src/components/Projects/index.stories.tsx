/* eslint-disable import/no-extraneous-dependencies */

import { storiesOf } from '@storybook/react';
import React from 'react';
import Projects from '.';
import projects from './__fixtures__';

storiesOf('Projects', module).add('default', () => <Projects projects={projects} />);
