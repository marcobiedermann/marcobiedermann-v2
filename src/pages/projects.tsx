import { PageProps } from 'gatsby';
import React from 'react';
import Grid from '../components/Grid';
import Layout from '../components/Layout';
import Projects from '../components/Projects';
import Section from '../components/Section';
import { useProjectsPage } from '../hooks/projects';

const BlogPage: React.FC<PageProps> = () => {
  const {
    allContentfulProject: { projects },
  } = useProjectsPage();

  return (
    <Layout>
      <Section>
        <Grid>
          <Projects projects={projects.map((project) => project.node)} />
        </Grid>
      </Section>
    </Layout>
  );
};

export default BlogPage;
