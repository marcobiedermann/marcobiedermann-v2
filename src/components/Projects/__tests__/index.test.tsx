/* eslint-disable import/no-extraneous-dependencies */

import React from 'react';
import renderer from 'react-test-renderer';
import Projects from '..';
import projects from '../__fixtures__';

describe('Projects component', () => {
  it('should render correctly', () => {
    const tree = renderer.create(<Projects projects={projects} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render empty projects correctly', () => {
    const tree = renderer.create(<Projects projects={[]} />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
