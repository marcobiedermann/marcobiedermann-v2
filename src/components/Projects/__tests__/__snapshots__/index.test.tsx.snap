// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projects component should render correctly 1`] = `
<DocumentFragment>
  <ol
    class=""
  >
    <li>
      <div
        id="6c5a5baa-8c8b-5580-8061-b1cad23712cc"
      >
        <figure>
          <a
            href="/projects/demo-project"
          >
            <figcaption>
              <div>
                <h3>
                  Demo Project
                </h3>
              </div>
            </figcaption>
          </a>
        </figure>
      </div>
    </li>
  </ol>
</DocumentFragment>
`;

exports[`Projects component should render empty projects correctly 1`] = `
<DocumentFragment>
  <ol
    class=""
  />
</DocumentFragment>
`;
